Ecms debugging

ScottSoren opened the pull request:
This PR contains a few bug fixes and feature enhancements, all related to EC and EC-MS measurements, in order to close a handful of Issues. 
See NEXT_CHANGES.rst for details.

I plan to release ixdat version 0.2.4 (which will also distribute Anna's gas_flux_calibration_curve method) once this is merged :)

ScottSoren commented on src/ixdat/readers/reading_tools.py:
```diff
@@ -149,9 +162,13 @@ def get_file_list(path_to_file_start=None, part=None, suffix=None):
-        folder = Path(path_to_file_start).parent
-        base_name = Path(path_to_file_start).name
-        file_list = [f for f in folder.iterdir() if f.name.startswith(base_name)]
+        path_to_file_start = Path(path_to_file_start)
+        if path_to_file_start.is_dir():
+            file_list = [f for f in path_to_file_start.iterdir() if f.is_file()]
+        else:
+            folder = path_to_file_start.parent
+            base_name = path_to_file_start.name
+            file_list = [f for f in folder.iterdir() if f.name.startswith(base_name)]
```
That's why there's the option to specify the suffix, which would get rid of README.md in the filter below.

ScottSoren commented on src/ixdat/readers/reading_tools.py:
```diff
@@ -11,6 +11,17 @@
 STANDARD_TIMESTAMP_FORM = "%d/%m/%Y %H:%M:%S"  # like '31/12/2020 23:59:59'
 USA_TIMESTAMP_FORM = "%m/%d/%Y %H:%M:%S"  # like '12/31/2020 23:59:59'
 FLOAT_MATCH = "[-]?\\d+[\\.]?\\d*(?:e[-]?\\d+)?"  # matches floats like '5' or '-2.3e5'
+DEFAULT_READER_NAMES = {
+    ".mpt": "biologic",
+    ".tsv": "zilien",
+    ".xrdml": "xrdml",
+    ".avg": "avantage",
+}
```
Two good points here. I've made these changes:
- Now if a suffix is provided without a "." at the start, the "." is added to match the syntax of `pathlib.Path.suffix`. So the user gets the same result when calling with `suffix="mpt"` as they do with `suffix=".mpt"`
- `Measurement.read_set` now raises a `ReadError` asking the user to check their inputs and/or files in case of an empty file list, rather than returning `None`.

matenestor replied:
Great!

ScottSoren commented on src/ixdat/readers/reading_tools.py:
```diff
@@ -11,6 +11,17 @@
+}
+
+
+def get_default_reader_name(path_to_file):
+    """Return a default reader if available given a file's full name with suffix"""
+    return DEFAULT_READER_NAMES.get(Path(path_to_file).suffix, None)
```
Fixed.

ScottSoren commented on src/ixdat/measurements.py:
```diff
@@ -1149,37 +1162,54 @@ def select_values(self, *args, **kwargs):
-                    "'{self.selector_name}' as a key-word argument"
-                )
-            kwargs[self.selector_name] = flat_args
+            if len(args) == 2 and isinstance(args[0], str):
+                # Then we can interpret the args as (selector_name, selector_value)
+                kwargs[args[0]] = args[1]
```
Fixed. `select_values("loop_number", 0, loop_number=1)` now raises a `ValueError`.

ScottSoren commented on src/ixdat/measurements.py:
```diff
@@ -205,6 +205,17 @@ def read(cls, path_to_file, reader, **kwargs):
             reader (str or Reader class): The (name of the) reader to read the file with.
             kwargs: key-word arguments are passed on to the reader's read() method.
         """
+        if not reader:
+            # Check if there is a default reader based on the file's suffix
+            from .readers.reading_tools import get_default_reader_name
+
+            reader = get_default_reader_name(path_to_file)
+            if not reader:
+                raise ValueError(
+                    f"There is no default reader for files of the type {path_to_file}. "
+                    "Please specify a reader to read this file."
+                )
+
```
Yes. I've now updated the dosctring to reflect this.

matenestor replied:
You also need to set the reader to None in the method parameters `reader=None`, so Python does not complain, when a user does not provide any.

AnnaWiniwarter commented on NEXT_CHANGES.rst:
```diff
@@ -20,8 +20,53 @@ techniques
+the case of no matching files.
+
+``Measurement.read`` (and by extension ``Measurement.read_set``) can now be called
+without a specified reader for certain known file types. To see which file types, use::
```
double ":" at the end of the line

ScottSoren replied:
This is intentional. It is a ReStructuredText syntax.
https://docutils.sourceforge.io/docs/user/rst/cheatsheet.txt

AnnaWiniwarter commented on NEXT_CHANGES.rst:
```diff
@@ -20,8 +20,53 @@ techniques
+  "cycle number".
+  Resolves `Issue #77 <https://github.com/ixdat/ixdat/issues/77>`_
+
+- ``MSInlet.gas_flux_calibration_curve`` added for multi-point calibration of MS data
```
This is already stated above in line 20 from my last PR, not sure if that was intentional?

ScottSoren replied:
Sorry, not intentional! Must have missed yours. I've deleted my duplicate.

AnnaWiniwarter commented on NEXT_CHANGES.rst:
```diff
@@ -20,8 +20,53 @@ techniques
+
+``ECMSPlotter.plot_measurment`` and ``plot_potential`` can take string options "all",
+"ms", and "ec" to specify the time span of the data to plot.
+Resolves `Issue #91 <https://github.com/ixdat/ixdat/issues/91>`_
```
cool! :)

AnnaWiniwarter reviewed (approved):
Looks good to me. Found a couple typos in NEXT_CHANGES.rst but no major issues.

ScottSoren commented on src/ixdat/measurements.py:
```diff
@@ -202,7 +202,9 @@ def read(cls, path_to_file, reader, **kwargs):
         Args:
             path_to_file (Path or str): The path to the file to read
-            reader (str or Reader class): The (name of the) reader to read the file with.
+            reader (str or Reader class): The (name of the) reader to read the file
+                with. If not specified, ixdat will try to determine the reader from the
+                file suffix.
```
fixed.

ScottSoren commented:
@matenestor , 
Brilliant comment with the ** notation in `select_values`. I've simplified the method by getting rid of the 2-argument syntex `select_values("series name", values)`, instead mentioning `select_values(**{"series name": values})` with an example in the docstring. The latter, as you point out, has the advantage of extension to multiple series, and is more pythonic.

I've also addressed the other comments.

Let me know if you think it looks good, and I'll merge and release 0.2.4 :D

matenestor reviewed (approved):
Looks great. I say it is ready to merge. :)
